File: config/libs/audio.lua
local ____lualib = require("lualib_bundle")
local __TS__StringEndsWith = ____lualib.__TS__StringEndsWith
local Error = ____lualib.Error
local RangeError = ____lualib.RangeError
local ReferenceError = ____lualib.ReferenceError
local SyntaxError = ____lualib.SyntaxError
local TypeError = ____lualib.TypeError
local URIError = ____lualib.URIError
local __TS__New = ____lualib.__TS__New
local ____exports = {}
function ____exports.audioPlugin(plugin, config)
    if config == nil then
        config = {}
    end
    config.plugin = plugin
    if not config.alias then
        if __TS__StringEndsWith(plugin, ".so") then
            error(
                __TS__New(Error, "Missing alias name. When providing a plugin path, an alias name must be provided."),
                0
            )
        end
        config.alias = plugin
    end
    return config
end
return ____exports

File: config/libs/audio.ts
import { writeFileSync } from 'fs';

const audioPluginList = new Set();

export function generateAudioMakefile(dest: string) {
    const content = `customBuild:
	make ${[...audioPluginList].join(' ')}
`;

    writeFileSync(dest, content);
}

export function audioPlugin(
    plugin: string,
    config: {
        // name alias of the plugin that will be used in the UI to interact with the plugin
        alias?: string;
        midiChannel?: number;
        midiCmd?: { parameter: string; cmd: string }[];
        // additional values to apply to the plugin
        [key: string]: unknown;
    } = {}
) {
    audioPluginList.add(plugin);
    config.plugin = plugin;
    if (!config.alias) {
        if (plugin.endsWith('.so')) {
            throw new Error(
                'Missing alias name. When providing a plugin path, an alias name must be provided.'
            );
        }
        config.alias = plugin;
    }
    return config;
}

export interface Track {
    id: number;
    plugins: ReturnType<typeof audioPlugin>[];
}
